### ISSUE TITLE
Gradle 8.0-RC1 Compatibility: Don't use internal Gradle classes

### ISSUE THREAD

**Issue by hfhbd** [JetBrains/gradle-intellij-plugin#1272]
**Describe the bug**
Use Intellij 1.11 with Gradle 8.0-rc-1
```stacktrace
Caused by: java.lang.NoSuchMethodError: 'void org.gradle.api.internal.artifacts.publish.ArchivePublishArtifact.<init>(org.gradle.api.tasks.bundling.AbstractArchiveTask)'
	at org.jetbrains.intellij.IntelliJPlugin$configureBuildPluginTask$1.invoke(IntelliJPlugin.kt:1383)
	at org.jetbrains.intellij.IntelliJPlugin$configureBuildPluginTask$1.invoke(IntelliJPlugin.kt:117)
	at org.jetbrains.intellij.IntelliJPlugin$inlined$sam$i$org_gradle_api_Action$0.execute(TaskContainerExtensions.kt)
	at org.gradle.api.internal.DefaultMutationGuard$1.execute(DefaultMutationGuard.java:45)
	at org.gradle.api.internal.DefaultMutationGuard$1.execute(DefaultMutationGuard.java:45)
	at org.gradle.configuration.internal.DefaultUserCodeApplicationContext$CurrentApplication$1.execute(DefaultUserCodeApplicationContext.java:123)
	at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1.run(DefaultCollectionCallbackActionDecorator.java:110)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:68)
	at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction.execute(DefaultCollectionCallbackActionDecorator.java:107)
	at org.gradle.internal.ImmutableActionSet$SetWithManyActions.execute(ImmutableActionSet.java:329)
	at org.gradle.api.internal.DefaultDomainObjectCollection.doAdd(DefaultDomainObjectCollection.java:262)
	at org.gradle.api.internal.DefaultNamedDomainObjectCollection.doAdd(DefaultNamedDomainObjectCollection.java:113)
	at org.gradle.api.internal.DefaultDomainObjectCollection.add(DefaultDomainObjectCollection.java:256)
	at org.gradle.api.internal.DefaultNamedDomainObjectCollection$AbstractDomainObjectCreatingProvider.tryCreate(DefaultNamedDomainObjectCollection.java:944)
	... 202 more
```
Reason: `ArchivePublishArtifact` is an internal Gradle classes and should not be used from plugin authors, because there is no binary compatibility guaranty. 

https://github.com/JetBrains/gradle-intellij-plugin/blob/ab1a88b94df74d1d905b09742b8ec6ad9ed86ecb/src/main/kotlin/org/jetbrains/intellij/IntelliJPlugin.kt#L1374

**To Reproduce**
Use Gradle 8.0-rc-1

**Expected behavior**
A working build

**Environment:**
 - OS: ubuntu 22.04
 - Gradle IntelliJ Plugin Version: 1.11.0
 - Gradle Version: 8.0-rc-1

**Additional context**
n/A

**Comment by hsz** [comment]
The issue is already fixed on the Gradle side and will be published with RC-2. Ref: [gradle/gradle#23440](https://github.com/gradle/gradle/issues/23440)

**Comment by hfhbd** [comment]
@hsz Are you sure? This is another class unrelated to this issue

**Comment by TWiStErRob** [comment]
That's not this issue, in my case I had no chance but to use internal classes as highlighted in https://github.com/gradle/gradle/issues/23431

You need to stop directly using internal classes:
https://github.com/JetBrains/gradle-intellij-plugin/blob/adae607148b18761d11b7912be54a1d6cce80601/src/main/kotlin/org/jetbrains/intellij/IntelliJPlugin.kt#L15-L16

https://github.com/JetBrains/gradle-intellij-plugin/blob/adae607148b18761d11b7912be54a1d6cce80601/src/main/kotlin/org/jetbrains/intellij/IntelliJPlugin.kt#L1291

https://github.com/JetBrains/gradle-intellij-plugin/blob/adae607148b18761d11b7912be54a1d6cce80601/src/main/kotlin/org/jetbrains/intellij/IntelliJPlugin.kt#L1293

I think this
https://github.com/JetBrains/gradle-intellij-plugin/blob/adae607148b18761d11b7912be54a1d6cce80601/src/main/kotlin/org/jetbrains/intellij/IntelliJPlugin.kt#L1290-L1296
can be replaced with something like:
```kotlin
project.artifacts {
    add(Dependency.ARCHIVES_CONFIGURATION, this@register)
}
project.components.add(IntelliJPluginLibrary())
```

**Comment by hsz** [comment]
@TWiStErRob  Thank you for your comment – I've fixed this internal API usage with 95892ddf9171053eedb4857c9c7fa31a0d19d662

As Gradle 8.0-RC.2 is already available, both issues should be resolved.

Can you please check the `1.12.1-SNAPSHOT` release?

**Comment by TWiStErRob** [comment]
Nice! I don't use the plugin myself, I'll let the others test the change.

Side question: What does addCandidate do? What is its observable effect?

**Comment by hfhbd** [comment]
I don't find `1.12.1-SNAPSHOT`: https://oss.sonatype.org/content/repositories/snapshots/org/jetbrains/intellij/plugins/gradle-intellij-plugin/

**Comment by hsz** [comment]
@hfhbd Thanks for pointing it out — I'll try to make it the right way.

@TWiStErRob `addCandidate` is supposed to inform the publishing-related tasks about the artifact which should be used in the publication process.

**Comment by TWiStErRob** [comment]
@hsz at first glance

**Comment by martinbonnin** [comment]
`1.12.1-SNAPSHOT` [works for us](https://github.com/apollographql/apollo-kotlin/pull/4644). Any chance it can get a proper `1.12.1` release so we can play a bit more with Gradle 8 release candidates?

**Comment by hfhbd** [comment]
@mgroth0 You have to use the plugin id as dependency: see https://github.com/JetBrains/gradle-intellij-plugin/issues/1272#issuecomment-1386676182
`implementation("org.jetbrains.intellij:org.jetbrains.intellij.gradle.plugin:1.12.2-SNAPSHOT")`

**Comment by hsz** [comment]
The issue is already fixed on both — the plugin and Gradle sides.
In addition, I've adjusted the snapshot publication, so the latest version is now available as
https://oss.sonatype.org/content/repositories/snapshots/org/jetbrains/intellij/plugins/gradle-intellij-plugin/1.12.3-SNAPSHOT/

**Comment by Eljah** [comment]
Having the same problem for

Unable to find method 'org.gradle.api.internal.artifacts.publish.ArchivePublishArtifact.<init>(Lorg/gradle/api/tasks/bundling/AbstractArchiveTask;)V'
org.gradle.api.internal.artifacts.publish.ArchivePublishArtifact.<init>(Lorg/gradle/api/tasks/bundling/AbstractArchiveTask;)V

Gradle 8.1.1 https://gradle.org/next-steps/?version=8.1.1&format=bin

and 221.5921,12 build of JetBrains gradle plugin

**Comment by hsz** [comment]
Update Gradle IntelliJ Plugin to the latest version.

**Comment by Eljah** [comment]
> Update Gradle IntelliJ Plugin to the latest version.

I can't, that us the last available

**Comment by hsz** [comment]
What is your setup?

**Comment by Eljah** [comment]
What exactly you need to know? Intellij idea: 
![image](https://github.com/JetBrains/gradle-intellij-plugin/assets/2749858/24fa2ae2-eff3-47e9-aca5-fc6c60911c51)

**Comment by TWiStErRob** [comment]
@Eljah Since we're in the `gradle-intellij-plugin` repo, the version from your Gradle build setup is the most important number. Your IDE version is not as important, since the error you're seeing is happening during a Gradle invocation (Build or Sync). Search for `org.jetbrains.intellij` (see [Usage section in docs](https://plugins.jetbrains.com/docs/intellij/tools-gradle-intellij-plugin.html#usage)) or `gradle-intellij-plugin` in your project files using "Find in Files" or command line grep. We're expecting that number to be lower than the latest release. If you read above in this issue, you can see this issue was reported against `1.11.0`, and fixed in `1.12.1-SNAPSHOT`. The snapshot got released as `1.13.0` in the end, and the latest release available is `1.14.0` (see https://github.com/JetBrains/gradle-intellij-plugin/releases).
